workers/chipset_device_worker/src/guestmem.rs

```diff
@@ -0,0 +1,207 @@
+                    GuestMemoryResponse::Write(result)
+                }
+                Err(RecvError::Closed) => {
+                    // The remote device may just drop guest memory if it
+                    // doesn't need it. We can just stop processing requests.
+                    break;
+                }
+                Err(RecvError::Error(e)) => {
+                    // TODO: Handle errors?
+                    panic!("guest memory proxy channel error: {:?}", e);
+                }
+            };
+            self.resp_send.send(response);
+        }
+    }
+}
+
+impl GuestMemoryRemoteBuilder {
+    /// Build the `GuestMemory` from the builder.
+    pub fn build(self, name: &str) -> GuestMemory {
+        GuestMemory::new(
+            name,
+            GuestMemoryRemote {
+                inner: Mutex::new(self.inner),
+            },
+        )
+    }
+}
+
+impl GuestMemoryRemote {
+    fn handle_blocking(&self, request: GuestMemoryRequest) -> GuestMemoryResponse {
+        let mut inner = self.inner.lock();
+        inner.req_send.send(request);
+        let resp = pal_async::local::block_on(inner.resp_recv.recv());
+        // TODO: Handle errors?
+        resp.unwrap()
+    }
+}
+
+// SAFETY: This implementation forwards all operations to the parent process,
+// which has the actual guest memory access. Since we don't have direct mapping
+// access in the worker process, we return None for mapping() and implement
+// all operations via fallback methods.
+unsafe impl GuestMemoryAccess for GuestMemoryRemote {
+    fn mapping(&self) -> Option<NonNull<u8>> {
+        None
+    }
+
+    fn page_fault(
+        &self,
+        _address: u64,
+        _len: usize,
+        _write: bool,
+        _bitmap_failure: bool,
+    ) -> guestmem::PageFaultAction {
+        guestmem::PageFaultAction::Fallback
+    }
+
+    fn max_address(&self) -> u64 {
+        // TODO: Is this the right value?
+        u64::MAX
+    }
+
+    unsafe fn read_fallback(
+        &self,
+        addr: u64,
+        dest: *mut u8,
+        len: usize,
+    ) -> Result<(), GuestMemoryBackingError> {
+        let GuestMemoryResponse::Read(data) =
+            self.handle_blocking(GuestMemoryRequest::Read { addr, len })
+        else {
+            unreachable!()
+        };
+        let data = data.map_err(|e| GuestMemoryBackingError::other(addr, e))?;
+        debug_assert_eq!(data.len(), len);
+        // SAFETY: caller guarantees dest is valid for write.
+        unsafe {
+            std::ptr::copy_nonoverlapping(data.as_ptr(), dest, len);
+        }
+        Ok(())
+    }
+
+    unsafe fn write_fallback(
+        &self,
+        addr: u64,
+        src: *const u8,
+        len: usize,
+    ) -> Result<(), GuestMemoryBackingError> {
+        // SAFETY: caller guarantees src is valid for read.
+        let data = unsafe { std::slice::from_raw_parts(src, len) }.to_vec();
```

Comment:
This isn't valid, the caller doesn't guarantee that `src` is immutable. Could point to volatile memory.